HTML API: Add `create_full_parser()` for non-fragment parsing.

dmsnell opened the pull request:
Trac ticket: Core-61576

## Description

As part of work to add more spec support to the HTML API, this patch adds support for the insertion modes from the initial start of a full document parse until IN BODY.

Modes after IN BODY are left to future work, but this change opens up the ability to start performing full document parses.

```diff
- Tests: 1525, Assertions: 3277, Skipped: 233.
+ Tests: 1519, Assertions: 3278, Skipped: 233.
```

<img width="867" alt="Screenshot 2024-07-05 at 9 41 50 PM" src="https://github.com/WordPress/wordpress-develop/assets/5431237/21d7c2a9-0354-41ba-8227-05ca7b2fd2dc">

dmsnell committed:
HTML API: Simplify breadcrumb accounting.

Since the HTML Processor started visiting all nodes in a document, both
real and virtual, the breadcrumb accounting became a bit complicated
and it's not entirely clear that it is fully reliable.

In this patch the breadcrumbs are rebuilt separately from the stack of
open elements in order to eliminate the problem of the stateful stack
interactions and the post-hoc event queue.

Breadcrumbs are greatly simplified as a result, and more verifiably
correct, in this construction.

dmsnell committed:
HTML API: Expand Unsupported class and make it available for debugging.

The HTML Processor internally throws an exception when it reaches HTML
that it knows it cannot process, but this exception is not made
available to calling code. It can be useful to extract more knowledge
about why it gave up, especially for debugging purposes.

In this patch, more context is added to the WP_HTML_Unsupported_Exception
and the last exception is made available to calling code, if it asks.

dmsnell committed:
HTML API: Introduce full parsing mode in HTML Processor.

The HTML Processor has only supported a specific kind of parsing mode
called _the fragment parsing mode_, where it behaves in the same way
that `node.innerHTML = html` does in the DOM. This mode assumes a
context node and doesn't support parsing an entire document.

This patch introduces the full parsing mode interface but leaves the
implementation incomplete, preparing the way for further work to add
that additional support.

See Core-61576

dmsnell committed:
HTML API: Add support for insertion modes before IN BODY.

As part of work to add more spec support to the HTML API, this patch adds
support for the insertion modes from the initial start of a full document
parse until IN BODY.

Modes after IN BODY are left to future work, but this change opens up the
ability to start performing full document parses.

See #61576.

dmsnell committed:
Merge branch 'trunk' into html-api/support-initial-mode

dmsnell committed:
Introduce bug to avoid WPCS rejecting the patch.

dmsnell committed:
Move unsupported elements into supported elements.

dmsnell committed:
Merge remote-tracking branch 'upstream/trunk' into html-api/support-initial-mode

dmsnell committed:
Merge branch 'trunk' into html-api/support-initial-mode

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -993,7 +1021,60 @@ public function get_current_depth(): int {
+			 */
+			case 'html':
+				$contents = $this->get_modifiable_text();
+				if ( ' html' !== $contents ) {
```
I see we have a todo comment following this, but I think this can be relaxed. I haven't found it in the specification, but browsers seem to match this case-insensitively. `HTML`, `HtMl` and `htmL` all seem to use no-quirks mode.

dmsnell replied:
this is fine, but I don't see it as a stumbling block at this point. the reason I arbitrarily limited it is that we don't parse DOCTYPE declarations properly. I'd kind of rather wait until that happens to start doing more here. that is, I don't want to write half a parser here and then a full parser in the Tag Processor.

maybe casing isn't that big of a deal. I'll think about it.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -993,7 +1021,60 @@ public function get_current_depth(): int {
+				 * > Then, switch the insertion mode to "before html".
+				 */
+				$this->state->insertion_mode = WP_HTML_Processor_State::INSERTION_MODE_BEFORE_HTML;
+				return true;
```
`next_token` does not seem to stop here, I'm unable to reach the doctype:

```php
$p = WP_HTML_Processor::create_full_parser('<!DOCTYPE html><body>'); $p->next_token();
echo $p->get_token_type() . ' ' . $p->get_tag();
// #tag HTML
```

dmsnell replied:
thanks. interestingly enough this is because we don't insert a node for the DOCTYPE declaration.

I'm not sure what to do either, because it's not accessible and it's not a child or parent of the root HTML node.

sirreal replied:
What do you mean by "it's not accessible?"

My initial thoughts are that this is an important part of the document and it should be reached via an interface like `next_token`.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1012,7 +1093,89 @@ private function step_initial(): bool {
+			if ( strlen( $text ) === strspn( $text, " \t\n\f\r" ) ) {
+				return $this->step();
+			}
+		}
```
In the spec this comes between comments and "html" start tag, any reason to move it here or could we respect the spec's order?

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1012,7 +1093,89 @@ private function step_initial(): bool {
+				/*
+				 * > Act as described in the "anything else" entry below.
+				 */
+				$is_excluded_closing_tag = true;
```
Rather than using flags, I like the goto approach we've used in other places to move directly to the referenced condition (this is also done in #7046).

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1031,7 +1194,90 @@ private function step_before_html(): bool {
+			case '-BODY':
+			case '-HTML':
+			case '-BR':
+				$is_excluded_closing_tag = true;
```
Same comment here regarding flags vs. goto.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1031,7 +1194,90 @@ private function step_before_html(): bool {
+			if ( strlen( $text ) === strspn( $text, " \t\n\f\r" ) ) {
+				return $this->step();
+			}
+		}
```
Why not do this in the switch?

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1012,7 +1093,89 @@ private function step_initial(): bool {
+		$html_element      = new WP_HTML_Token( $new_bookmark, 'HTML', false );
+		$this->state->stack_of_open_elements->push( $html_element );
+		$this->state->insertion_mode = WP_HTML_Processor_State::INSERTION_MODE_BEFORE_HEAD;
+		return $this->step( self::REPROCESS_CURRENT_NODE );
```
Should this be using the `insert_virtual_node` method?

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1031,7 +1194,90 @@ private function step_before_html(): bool {
+		$head_element      = new WP_HTML_Token( $new_bookmark, 'HEAD', false );
+		$this->state->stack_of_open_elements->push( $head_element );
+		$this->state->head_element   = $head_element;
+		$this->state->insertion_mode = WP_HTML_Processor_State::INSERTION_MODE_IN_HEAD;
```
Should this be using the `insert_virtual_node` method?

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1050,7 +1296,177 @@ private function step_before_head(): bool {
+			}
+		}
+
+		$is_excluded_closing_tag = false;
```
Same comments from other sections - I'd like to follow the goto pattern we're using elsewhere over this flag, and could we handle text nodes in the main switch?

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1050,7 +1296,177 @@ private function step_before_head(): bool {
+			case '+META':
+				$this->insert_html_element( $this->state->current_token );
+
+				/*
+				 * > If the active speculative HTML parser is null, then:
+				 * >   - If the element has a charset attribute, and getting an encoding from
+				 * >     its value results in an encoding, and the confidence is currently
+				 * >     tentative, then change the encoding to the resulting encoding.
+				 */
+				$charset = $this->get_attribute( 'charset' );
+				if ( is_string( $charset ) && 'tentative' === $this->state->encoding_confidence ) {
+					$this->bail( 'Cannot yet process META tags with charset to determine encoding.' );
+				}
+
+				/*
+				 * >   - Otherwise, if the element has an http-equiv attribute whose value is
+				 * >     an ASCII case-insensitive match for the string "Content-Type", and
+				 * >     the element has a content attribute, and applying the algorithm for
+				 * >     extracting a character encoding from a meta element to that attribute's
+				 * >     value returns an encoding, and the confidence is currently tentative,
+				 * >     then change the encoding to the extracted encoding.
+				 */
+				$http_equiv = $this->get_attribute( 'http-equiv' );
+				$content    = $this->get_attribute( 'content' );
+				if (
+					is_string( $http_equiv ) &&
+					is_string( $content ) &&
+					0 === strcasecmp( $http_equiv, 'Content-Type' ) &&
+					'tentative' === $this->state->encoding_confidence
+				) {
+					$this->bail( 'Cannot yet process META tags with http-equiv Content-Type to determine encoding.' );
+				}
```
To be clear, we can't hit this condition now because the parser will not start without a certain (utf8) encoding. Correct?

**I'll correct myself** - we do hit this condition in the fragment parser. I suspect the fragment parser should also set the encoding and encoding confidence.

---

Both of these cases depend on confidence being `"tentative"`, maybe we could put that condition up front and skip any additional processing.

dmsnell replied:
in the fragment parser I think we can set the encoding confidence to `certain` based on the provided argument to `create_fragment()`. _if_ we ever support tentative confidence we can update this then.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1050,7 +1296,177 @@ private function step_before_head(): bool {
+			 */
+			case '+TEMPLATE':
+			case '-TEMPLATE':
+				$this->bail( 'Cannot process TEMPLATE elements in the HEAD.' );
```
This is omitted because #7046 handles it, correct?

dmsnell replied:
yes

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1069,7 +1485,99 @@ private function step_in_head(): bool {
+			}
+		}
+
+		$is_excluded_closing_tag = false;
```
Same - I'd prefer goto over flag and It's not clear to me why the `#text` isn't handled in a single switch in this case.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -340,6 +340,34 @@ public static function create_fragment( $html, $context = '<body>', $encoding =
+		}
+
+		$processor                             = new static( $html, self::CONSTRUCTOR_UNLOCK_CODE );
+		$processor->state->encoding            = $known_definite_encoding;
+		$processor->state->encoding_confidence = 'certain';
```
I think we need to set these in the fragment parser as well.

dmsnell replied:
right now the fragment parser _only_ supports `UTF-8`, which would be `certain` confidence. because that's inferred from the parent document, I don't feel like we need to add more there yet. it's still my intention _only_ to support UTF-8, as I would prefer we get in the habit of converting before processing.

thoughts?

sirreal replied:
UTF-8 only support seems like the right choice at this time.

It's good that we set the encoding and encoding confidence in the fragment parser. Before that change it was resulting in different behavior for `<meta charset="anything">` tags between the full and fragment parsers. That's fixed by always setting `certain` encoding confidence.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1088,7 +1596,141 @@ private function step_in_head_noscript(): bool {
+		$op_sigil   = '#tag' === $token_type ? ( $is_closer ? '-' : '+' ) : '';
+		$op         = "{$op_sigil}{$token_name}";
+
+		/*
+		 * > A character token that is one of U+0009 CHARACTER TABULATION,
+		 * > U+000A LINE FEED (LF), U+000C FORM FEED (FF),
+		 * > U+000D CARRIAGE RETURN (CR), or U+0020 SPACE
+		 */
+		if ( '#text' === $op ) {
+			$text = $this->get_modifiable_text();
+			if ( strlen( $text ) === strspn( $text, " \t\n\f\r" ) ) {
+				// Insert the character.
+				$this->insert_html_element( $this->state->current_token );
+				return true;
+			}
+		}
+
+		$is_excluded_closing_tag = false;
```
Same - I'd prefer goto over flag and could we handle `#text` in the switch?

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1069,7 +1485,99 @@ private function step_in_head(): bool {
+			case '+STYLE':
+				return $this->step_in_head();
+
+			/*
+			 * > An end tag whose tag name is "br"
+			 *
+			 * This should never happen, as the Tag Processor prevents showing a BR closing tag.
+			 */
+			case '-BR':
+				/*
+				 * > Act as described in the "anything else" entry below.
+				 */
+				$is_excluded_closing_tag = true;
+				break;
```
I didn't understand why `</br>` would get special handling here, but it's to prevent the end tag from being ignored. `</br>` has special handling to be treated as an opener.

It is handled correctly in this PR 👍

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1088,7 +1596,141 @@ private function step_in_head_noscript(): bool {
+		 * > Anything else
+		 * > Insert an HTML element for a "body" start tag token with no attributes.
+		 */
+		$currently_at      = $this->bookmarks[ $this->state->current_token->bookmark_name ];
+		$new_bookmark      = $this->bookmark_token();
+		$this->bookmarks[] = new WP_HTML_Span( $currently_at->start + $currently_at->length, 0 );
+		$this->state->stack_of_open_elements->push(
+			new WP_HTML_Token(
+				$new_bookmark,
+				'BODY',
+				false
+			)
+		);
```
Should this use the `insert_virtual_node` method?

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -4319,6 +4961,53 @@ public static function is_void( $tag_name ): bool {
+	 * @param string $label A string which may specify a known encoding.
+	 * @return string|null Known encoding if matched, otherwise null.
+	 */
+	protected static function get_encoding( string $label ): ?string {
```
This looks correct but seems to be unused in this PR, was that intentional?

dmsnell replied:
it was intentional, but maybe it's not necessary. if we support inferring the character encoding it would be useful.

sirreal reviewed (commented):
Nice, there's a lot in this one! I've left many comments in questions.

One thing I noticed is that it seems like _any document_ proceeds through a number of insertion modes that ultimately creates `HTML > HEAD > BODY` (even if empty). This PR does not seem to do that.

---

I would like to figure out the html5lib tests to get more coverage here, although it's fine to do that after this lands if you'd prefer. We'd need to disable this flag and change the processor (fragment or full) depending on each test case. It will require more changes to fix the tree representation.

dmsnell committed:
Merge branch 'trunk' into html-api/support-initial-mode

dmsnell committed:
Replace separated text handling with gotos

dmsnell committed:
Use insert_virtual_node() instead of directly inserting virtual nodes.

dmsnell committed:
WPCS

sirreal committed:
Order before html cases according to spec

The spec handles doctype-comment-text, adjust
implementation to match that order.

sirreal reviewed (approved):
This seems ready to land.

- I would like to expose the doctype token somehow.
- I'd also like to figure out the full html processor and html5lib-tests.

These are related and can be addressed in a follow-up.

dmsnell committed:
Merge branch 'trunk' into html-api/support-initial-mode

dmsnell commented:
Merged in [[58836]](https://core.trac.wordpress.org/changeset/58836)
https://github.com/wordpress/wordpress-develop/commit/883146e8a5b79f3accf6a171a6c53a4716a2b61c
